teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+
+Therefore, a built-in mechanism is desired to allow `Task`/`Pipeline` authors to declare upfront a list of allowed values for a parameter at authoring time. 
+
+In addition, this feature will enable Tekton to become a step closer to [SLSA hermeticity requirement](https://slsa.dev/spec/v0.1/requirements#hermetic), which requires that all inputs to a build must be fully declared up front. 
+
+
+### Goals
+- Design a Tekton built-in `param` input validation mechanism that constrains the user-provided `string param` value to a set of allowed constants predefined by the author.
+
+### Non-Goals
+- Meeting [SLSA hermeticity](https://slsa.dev/spec/v0.1/requirements#hermetic) requirement.
+
+### Use Cases  
+- As the [buildah](https://github.com/tektoncd/catalog/tree/main/task/buildah/0.5) Task Author, I want to specify the valid values (`oci` or `docker`) of the `param` [`FORMAT`](https://github.com/tektoncd/catalog/blob/4754df9c260e8f102616170175cb132cfc3c5544/task/buildah/0.5/buildah.yaml#L42C1-L44C19). I want to fail the validation if the input is **NOT** valid.
+
+  Without input validation, the buildah task will fail at `pod` runtime due to the invalid `FORMAT` value. The error message is hidden in the container log:
+  
+    ```yaml
+    apiVersion: tekton.dev/v1
+    kind: TaskRun
+    name: buildah-run
+    spec:
+      taskRef:
+        name: buildah
+      params:
+        - name: FORMAT
+          value: "invalid-val"
+      ...
+    ```
+    
+    ```bash
+    # the execution of the above task fails with log:
+    $ kubectl logs buildah-run-pod -c step-build
+    unrecognized image type "invalid-val"
+    ```
+    
+  As a workaround, the `Task` Author must add a validation step before building the image using buildah:
+
+  ```yaml
+  apiVersion: tekton.dev/v1beta1
+  kind: Task
+  metadata:
+    name: buildah
+  spec:
+    params:
+    - name: FORMAT
+      description: The format of the built container, oci or docker
+      default: "oci"
+    ....
+    steps:
+      - name: validate-format-param
+        image: bash
+        script: |
+          # a script to check if $(params.FORMAT) is either `oci` or `docker`.
+      - name: build 
+        image: quay.io/buildah/stable:v1.23.3
+        script: |
+          # build image using buildah
+        ...
+  ```
+
+- As the [gcs-create-bucket](https://github.com/tektoncd/catalog/tree/main/task/gcs-create-bucket/0.1) `Task` Author, I want to specify the valid values (`STANDARD`, `NEARLINE`, `COLDLINE`, or `ARCHIVE` ) of `param` [`storageClass`](https://github.com/tektoncd/catalog/blob/4754df9c260e8f102616170175cb132cfc3c5544/task/gcs-create-bucket/0.1/gcs-create-bucket.yaml#L31C1-L35C22). I want to fail the validation if the input is NOT valid.
+
+- As a `Task`/`Pipeline` Author in an organization, I want to enhance the Task security by declaring upfront the allowed/approved images digests that can be used in the `Task`.
```

Comment:
Regarding security, wondering if enum is most effective with remote Task/Pipelines since for local Task/Pipelines the enums could be changed regardlessly?

Reply:
I think it can enhance both remote and local resources. Even with local resources, users can inject any random strings (code snippts, sql queries) if the param is not bounded, and the users do not necessarily need access to modify the Task.

Reply:
Makes sense, thanks for the clarifications @QuanZhang-William